libs/test-utils/src/mock_function.ts

```diff
@@ -0,0 +1,180 @@
+}
+
+interface AsyncReturnHelpers<Func extends AnyFunc> {
+  resolves(output: Awaited<ReturnType<Func>>): void;
+  throws(error: unknown): void;
+}
+
+export interface MockFunction<Func extends AnyFunc> {
+  (...input: Parameters<Func>): ReturnType<Func>;
+  expectCallWith(
+    ...input: Parameters<Func>
+  ): ReturnType<Func> extends Promise<unknown>
+    ? AsyncReturnHelpers<Func>
+    : ReturnHelpers<Func>;
+  reset(): void;
+  assertComplete(): void;
+}
+
+interface MockFunctionState<Func extends AnyFunc> {
+  expectedCalls: Array<ExpectedCall<Func>>;
+  actualCalls: Array<Call<Func>>;
+}
+
+function formatFunctionCall(name: string, input: unknown[]): string {
+  return `${name}(${input
+    .map((arg) => inspect(arg, { depth: null }))
+    .join(', ')})`;
+}
+
+/**
+ * Creates a mock function, similar to jest.fn(), but with stricter rules to
+ * help you avoid making unintentional mistakes.
+ *
+ * Example usage:
+ *
+ *  // Let's say we have a real function, add, that we want to mock
+ *  function add(num1: number, num2: number): number {
+ *    return num1 + num2;
+ *  }
+ *
+ *  // We can create a mock function that behaves the same way
+ *  const addMock = mockFunction<typeof add>('add');
+ *  addMock.expectCallWith(1, 2).returns(3);
+ *  addMock(1, 2); // -> returns 3
+ *  addMock.expectCallWith(1, NaN).throws(new Error('Cant add NaN'));
+ *  addMock(1, NaN); // -> throws Error('Cant add NaN')
+ *
+ *  // Unlike jest.fn(), mockFunction is strict...
+ *
+ *  // Each expected call can only be used once, in order
+ *  addMock.expectCallWith(1, 2).returns(3);
+ *  addMock(1, 2); // -> returns 3
+ *  addMock(1, 2); // -> throws an error because we didn't expect this call
+ *
+ *  // The actual input must match the expected input
+ *  addMock.expectCallWith(1, 2).returns(3);
+ *  addMock(1, 3); // -> throws an error because the input doesn't match
+ *
+ *  // Every expected call must be used
+ *  addMock.expectCallWith(1, 2).returns(3);
+ *  addMock.assertComplete(); // -> throws an error because we didn't use the expected call
+ *
+ * Recommendations:
+ * - Always call mockFunction.assertComplete() to ensure that all expected calls were used
+ * - If you are going to reuse a mock between test cases, call mockFunction.reset() in afterEach
+ */
+// eslint-disable-next-line vx/gts-no-return-type-only-generics
+export function mockFunction<Func extends AnyFunc>(
+  name: string
+): MockFunction<Func> {
+  const state: MockFunctionState<Func> = {
+    expectedCalls: [],
+    actualCalls: [],
+  };
+
+  const mock: MockFunction<Func> = (...input: Parameters<Func>) => {
+    state.actualCalls.push({ input });
+    const expectedCall = state.expectedCalls[state.actualCalls.length - 1];
+    if (!expectedCall) {
+      const message = `Unexpected call to mock function: ${formatFunctionCall(
+        name,
+        input
+      )}`;
+      // Log to console since exceptions can be swallowed depending on how the mock function is used
```

Comment:
I know this means that, in some cases, we'll double log, but I see why this is necessary 👍

Reply:
I decided to move this exception trapping functionality to the mock grout client, since the main reason we need it is that we have a bad habit of not wrapping api call promises in try/catch. I don't think it makes as much sense for a general mock function to not raise an exception. b0047f544